### PR TITLE
Original NKN kernel ( Flux version )

### PR DESCRIPTION
**Pull request by HamletWantToCode**
Hi @willtebbutt , I receive an error when rebasing PR #92 , it said "A pull request already exists for base branch 'master' and head branch 'master'". So I checkout my original code and open a PR again, this should be the same as PR #92 .

### CODE REVIEW

File: src/Stheno.jl [hunk 1]
```diff
@@ -41,6 +41,7 @@ module Stheno
     # Atomic GP objects.
     include(joinpath("gp", "mean.jl"))
     include(joinpath("gp", "kernel.jl"))
+		include(joinpath("gp", "neural_kernel_network.jl"))
```
Comment (willtebbutt) [inline src/Stheno.jl]:
```suggestion
    include(joinpath("gp", "neural_kernel_network.jl"))
```

File: src/gp/kernel.jl [hunk 2]
```diff
@@ -106,15 +108,20 @@ The usual periodic kernel derived by mapping the input domain onto the unit circ
...
 # Binary methods.
-ew(k::PerEQ, x::AV{<:Real}, x′::AV{<:Real}) = exp.(.-2 .* sin.(π .* abs.(x .- x′)).^2)
-pw(k::PerEQ, x::AV{<:Real}, x′::AV{<:Real}) = exp.(.-2 .* sin.(π .* abs.(x .- x′')).^2)
+ew(k::PerEQ, x::AV, x′::AV) = _pereq.(ew(Euclidean(), x, x′), k.logl[1]) 
+pw(k::PerEQ, x::AV, x′::AV) = _pereq.(pw(Euclidean(), x, x′), k.logl[1])
 
 # Unary methods.
-ew(::PerEQ, x::AV{<:Real}) = ones(eltype(x), length(x))
-pw(k::PerEQ, x::AV{<:Real}) = pw(k, x, x)
+ew(k::PerEQ, x::AV) = _pereq.(ew(Euclidean(), x), k.logl[1])
```
Comment (willtebbutt) [inline src/gp/kernel.jl]: Why has this been changed?

File: src/gp/neural_kernel_network.jl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ew(nkn::NeuralKernelNetwork, x, x′) = _rebuild_diag(nkn.chain(ew(nkn.player, x, x′)))
+pw(nkn::NeuralKernelNetwork, x, x′) = _rebuild_kernel(nkn.chain(pw(nkn.player, x, x′)), length(x), length(x′))
+
+function Base.show(io::IO, kernel::NeuralKernelNetwork)
+    print(io, "NeuralKernelNetwork(")
+    join(io, [kernel.player, kernel.chain], ", ")
+    print(io, ")")
+end
+
+
```
Comment (willtebbutt) [inline src/gp/neural_kernel_network.jl]:
```suggestion
```

File: src/gp/kernel.jl (hunk 2 above)
Comment (HamletWantToCode) [reply src/gp/kernel.jl]: The `PerEQ` should have three parameters: `σ²` ( scaling factor ), `l` ( length scale ) & `p` ( period ), using `Scaled` & `Stretched` can specify the first two, but I don't find way for the third one ( maybe Stheno already has it implemented, please indicate to me if I miss it ). 

As is mentioned [here](https://www.cs.toronto.edu/~duvenaud/cookbook/), the formula for periodic kernel looks like ( without scaling & length factor ):
```julia
exp(-2.0 * sin(π * d)^2 / l^2)
```

While Stheno's current implementation is:
```julia
exp.(.-2 .* sin.(π .* abs.(x .- x′)).^2)
```
and one can only pass in `AbstractVector` type to  `ew` & `pw` method for `PerEQ`, when pass in `ColVecs`, I got the following error message:
```julia
julia> pereq = Stheno.PerEQ()
julia> x = rand(3, 10);
julia> pw(pereq, ColVecs(x), ColVecs(x))
ERROR: MethodError: no method matching pairwise(::Stheno.PerEQ, ::ColVecs{Float64,Array{Float64,2}}, ::ColVecs{Float64,Array{Float64,2}})
```
also, Stheno doesn't explicitly export the `PerEQ`.

> I'm using Stheno v0.6.1 with julia v1.3.1 under macOS

**Comment by HamletWantToCode** [comment]
Hi @willtebbutt:

> if you could use `Requires.jl`

`Requires.jl` is used, I have also added [`add_functor.jl`](https://github.com/HamletWantToCode/Stheno.jl/blob/nkn_kernel/src/gp/add_functor.jl), which automatically generate calls e.g. `@functor EQ` when `Flux` is used.

> Could you please hook the NKN into the standard testing infrastructure?

NKN has now be tested.

## To be discussed
1. **How to add constraints to kernel's parameters without affecting `Stheno`'s internal implementation ?** Parameters like `σ²` ( scaling factor ) & `l` ( length scale ) are required to be positive during optimization. In PR #94 , I resolved it by adding an additional field `f` ( e.g. `f = x->exp(x)` ) to kernels, and parameters will first be transformed from `x` to `f(x)` and then passed to `ew` or `pw` functions. I think there should be some cleaner methods for this problem ( I'm seeking something behaves as `Scaled` & `Stretched` types, maybe we can define a `Constraint` type for it ).

File: src/Stheno.jl [hunk 4]
```diff
@@ -42,7 +42,7 @@ module Stheno
     include(joinpath("gp", "mean.jl"))
     include(joinpath("gp", "kernel.jl"))
     include(joinpath("gp", "gp.jl"))
-
+
```
Comment (willtebbutt) [inline src/Stheno.jl]:
```suggestion
```

File: src/gp/neural_kernel_network.jl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+# Neural Kernel Network, since kernel space ( stationary kernel ) is closed under linear combination 
+# ( with positive coefficient ) and element-wise multiplication, we can use a neural network like structure
+# to build composite kernels. This type contains a `Primitive` layer which holds basic kerenls and a specialised
+# nerual network architecture to perform kernel composition. It should function like a normal `Stheno` kernel.
+struct NeuralKernelNetwork{PT, CT} <: Kernel
+    player::PT
```
Comment (willtebbutt) [inline src/gp/neural_kernel_network.jl]: Please consider renaming to `primitive_layer` or simply `primitives`.

File: test/gp/kernel.jl [hunk 6]
```diff
@@ -192,6 +194,19 @@ using LinearAlgebra
...
+            k1 = 0.5 * stretch(EQ(), 0.1)
+            k2 = 1.0 * stretch(PerEQ(1.0), 0.2)
+            prim_layer = Primitive(k1, k2)
+
+            lin = LinearLayer(2, 2)
+            nn = Chain(lin, product)
+
+            nkn = NeuralKernelNetwork(prim_layer, nn)
+            differentiable_kernel_tests(nkn, ȳ, Ȳ, Ȳ_sq, X0, X1, X2; atol=1e-7, rtol=1e-7)
+        end
```
Comment (willtebbutt) [inline test/gp/kernel.jl]: This is a great start. Would be good to implement a few tests to make sure that the `NKN` recovers other kernels when it's meant to. eg. a `GP` with an `NKN` kernel with only a linear layer should be equivalent to a `GP` with a particular composite kernel. Having one or two tests that check that this holds (by, say, checking that the covariance matrices produced in both cases are almost equal) would be very helpful.

**Comment by willtebbutt** [comment]
> How to add constraints to kernel's parameters without affecting Stheno's internal implementation ? Parameters like σ² ( scaling factor ) & l ( length scale ) are required to be positive during optimization. In PR #94 , I resolved it by adding an additional field f ( e.g. f = x->exp(x) ) to kernels, and parameters will first be transformed from x to f(x) and then passed to ew or pw functions. I think there should be some cleaner methods for this problem ( I'm seeking something behaves as Scaled & Stretched types, maybe we can define a Constraint type for it ).

I agree that we need to resolve this. My concern is that this is quite hard to do. I suspect that the solution you've come up with in #94 is the best that we can do for now, so it probably makes sense to go with that for now. Probably we'll need to refine the solution a bit in future PRs, but that approach probably makes sense for now.

**Comment by HamletWantToCode** [comment]
Hi @willtebbutt , just finish adding extra test cases for NKN and add constraints to kernels.

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+# construct kernels
+iso_lin_kernel1 = stretch(Linear(), [0.0])
+iso_per_kernel1 = [log(1.0)] * stretch(PerEQ([log(l)]), [log(l)])
+iso_eq_kernel1 = [log(1.0)] * stretch(EQ(), [log(l/4.0)])
+iso_rq_kernel1 = [log(1.0)] * stretch(RQ([log(0.2)]), [log(2.0*l)])
+iso_lin_kernel2 = stretch(Linear(), [0.0])
+iso_rq_kernel2 = [log(1.0)] * stretch(RQ([log(0.1)]), [log(l)])
+iso_eq_kernel2 = [log(1.0)] * stretch(EQ(), [log(l)])
+iso_per_kernel2 = [log(1.0)] * stretch(PerEQ([log(l/4.0)]), [log(l/4.0)])
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]: Is it necessary to have variances for these kernels? Won't the first linear layer handle that?

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+
+# sum product network
+linear1 = LinearLayer(8, 8)
+linear2 = LinearLayer(4, 4)
+linear3 = LinearLayer(2, 1)
+
+# NKN
+player = Primitive(iso_lin_kernel1, iso_per_kernel1, iso_eq_kernel1, iso_rq_kernel1,
+                     iso_lin_kernel2, iso_rq_kernel2, iso_eq_kernel2, iso_per_kernel2)
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]: Please consider re-writing as:
```suggestion
player = Primitive(
    stretch(Linear(), [0.0]),
    stretch(PerEQ([log(l)]), [log(l)]),
    stretch(EQ(), [log(l/4.0)]),
    stretch(RQ([log(0.2)]), [log(2.0*l)]),
    stretch(Linear(), [0.0]),
    stretch(RQ([log(0.1)]), [log(l)]),
    stretch(EQ(), [log(l)]),
    stretch(PerEQ([log(l/4.0)]), [log(l/4.0)]),
)
```
and removing the kernel creation above. They don't get used again, so no need to give them names imho

File: test/gp/kernel.jl [hunk 9]
```diff
@@ -196,15 +196,41 @@ using LinearAlgebra
...
+            @timedtestset "kernel composition test" begin
+                k1 = 0.5 * stretch(EQ(), 0.1)
+                k2 = 1.0 * stretch(PerEQ(1.0), 0.2)
+                prim_layer = Primitive(k1, k2)
+                lin = LinearLayer(ones(1, 2))
+                nkn_add_kernel = NeuralKernelNetwork(prim_layer, lin)
+                nkn_prod_kernel = NeuralKernelNetwork(prim_layer, product)
+
+                sum_k = Stheno.softplus(1.0)*k1 + Stheno.softplus(1.0)*k2
+                prod_k = k1 * k2
```
[the remark this answers is not in the document]
Comment (HamletWantToCode) [reply test/gp/kernel.jl]: Yes, the `softplus` function is used to make sure matrix elements of linear transformation always be positive. See [here](https://github.com/HamletWantToCode/Stheno.jl/blob/1f85c9a20ff4074c81482832ed2b7b6e4117c0e5/src/gp/neural_kernel_network.jl#L10-L16).

Therefore, in `lin = LinearLayer(ones(1, 2))`,  the actual weights used for calculation aren't `ones(1, 2)`, it should be `softplus.(ones(1, 2))`.

**Comment by HamletWantToCode** [comment]
Hi @willtebbutt , thank you for your advise, they are really helpful. Example is restructure based on current implementation, also a README.md is added.

File: examples/flux_integration/neural_kernel_network/README.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# Neural kernel network
+When apply neural network to data, we are essentially superpose different features of data and pass the result to a nonlinear
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/README.md]:
```suggestion
When applying a neural network to data, we compute different linear features of data and pass the result to a nonlinear
```

File: examples/flux_integration/neural_kernel_network/README.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+activation function. Similar things can be done to kernel functions ( stationary kernels ) as well, and combination of the kernel functions follows
+the rules ( $k_1,\,k_2$ are kernel functions ):
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/README.md]:
```suggestion
activation function. A similar thing can be done with kernels as well. In particular, kernels can be combined according to the following rules ( $k_1,\,k_2$ are kernel functions ):
```

File: examples/flux_integration/neural_kernel_network/README.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+1. For $\lambda_1,\,\lambda_2\in\mathbb{R}^+$, $\lambda_1 k_1+\lambda_2 k_2$ is also a valid kernel function
+
+2. The product $k_1k_2$ is a kernel function
+
+Therefore, we can introduce a neural network like structure, which consists superposition, product and nonlinear activation operations ( similar to the idea fo sum and product network ). Acting such a network on top of some basic kernels, e.g. exponential quardratic kernel, periodic kernel, linear kernel, etc., will results in a new kernel function which is able to extract more complicated features. In addition, the neural kernel network is differentiable w.r.t it's parameters, which enables us to determine it via gradient based optimization methods. This also illuminates previous black-box kernel formation process.
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/README.md]:
```suggestion
We can use these to specify a neural network like structure, which consists of linear transformations, products, and nonlinear activation operations (similar to Sum-Product networks). Such a structure is called the Neural Kernel Network (NKN) [1]. Building such a network on top of some basic kernels, e.g. exponentiated-quadratic, periodic, linear, etc., will result in a new kernel function which is able to extract more complicated features. In addition, the neural kernel network is differentiable w.r.t it's parameters, which enables these to be found via gradient based optimisation methods. This is distinct from previous compositional kernel search methods eg. [2], which search over an unbounded space of possible kernel structures.
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 13]
```diff
@@ -112,12 +112,15 @@ function predict(X, Xtrain, ytrain)
 end
 
 posterior = predict(Year, Xtrain, ytrain)
-pred_y = mean(posterior)
+post_dist = marginals(posterior)
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
post_marginals = marginals(posterior)
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 14]
```diff
@@ -112,12 +112,15 @@ function predict(X, Xtrain, ytrain)
...
+pred_y = mean.(post_dist)
+var_y = std.(post_dist)
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
pred_y = mean.(post_marginals)
var_y = std.(post_marginals)
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 15]
```diff
@@ -112,12 +112,15 @@ function predict(X, Xtrain, ytrain)
...
 pred_oy = @. pred_y*ytrain_std+ytrain_mean
+pred_oσ = @. var_y*ytrain_std
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
pred_oy = @. pred_y * ytrain_std + ytrain_mean
pred_oσ = @. var_y * ytrain_std
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 16]
```diff
@@ -112,12 +112,15 @@ function predict(X, Xtrain, ytrain)
...
 
 plt = plot(xlabel="Year", ylabel="Airline Passenger number", legend=true)
-plot!(plt, year, pred_oy, title="Time series prediction",label="95% predictive confidence region")
+plot!(plt, year, pred_oy, ribbons=3*pred_oσ, title="Time series prediction",label="95% predictive confidence region")
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
plot!(plt, year, pred_oy;
    ribbons=3*pred_oσ,
    title="Time series prediction",
    label="95% predictive confidence region",
)
```

File: test/gp/kernel.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (willtebbutt) [reply test/gp/kernel.jl]: I think it would still be better to use randomised weights. Having the same values for each of the elements is still a slightly weird edge case. As a rule, I try to make any quantities used in tests as un-special as (unless specifically testing potentially worrisome corner cases) possible to avoid potential edge cases -- randomisation is a good way to achieve this.

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 17]
```diff
@@ -54,28 +54,28 @@ end
...
 
 # NKN
-player = Primitive(iso_lin_kernel1, iso_per_kernel1, iso_eq_kernel1, iso_rq_kernel1,
-                     iso_lin_kernel2, iso_rq_kernel2, iso_eq_kernel2, iso_per_kernel2)
-nn = Chain(linear1, Product, linear2, Product, linear3)
-nkn = NeuralKernelNetwork(player, nn)
+# NOTE: the numerical values are in log-scale
+# therefore, stretch(Linear(), 0.0, exp) ≡ stretch(Linear(), 1.0)
+# `exp` is a constraint added to the kernel parameters to maintain
+# their positiveness during the computation
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
# their positiveness during the optimisation.
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 18]
```diff
@@ -54,28 +54,28 @@ end
...
+    stretch(Linear(), 0.0, x->exp(-x)),
+    stretch(PerEQ(log(l), exp), log(l), x->exp(-x)),
+    stretch(EQ(), log(l/4.0), x->exp(-x)),
+    stretch(RQ(log(0.2), exp), log(2.0*l), x->exp(-x)),
+    stretch(Linear(), 0.0, x->exp(-x)),
+    stretch(RQ(log(0.1), exp), log(l), x->exp(-x)),
+    stretch(EQ(), log(l), x->exp(-x)),
+    stretch(PerEQ(log(l/4.0), exp), log(l/4.0), x->exp(-x))
+)
+nn = Chain(linear1, product, linear2, product, linear3)
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
nn = Chain(
    LinearLayer(8, 8),
    product,
    LinearLayer(4, 4),
    product,
    LinearLayer(2, 1),
)
```

File: examples/flux_integration/neural_kernel_network/time_series.jl [hunk 19]
```diff
@@ -54,28 +54,28 @@ end
...
-iso_rq_kernel2 = [log(1.0)] * stretch(RQ([log(0.1)]), [log(l)])
-iso_eq_kernel2 = [log(1.0)] * stretch(EQ(), [log(l)])
-iso_per_kernel2 = [log(1.0)] * stretch(PerEQ([log(l/4.0)]), [log(l/4.0)])
-
-
-# sum product network
+# neural network
 linear1 = LinearLayer(8, 8)
 linear2 = LinearLayer(4, 4)
 linear3 = LinearLayer(2, 1)
```
Comment (willtebbutt) [inline examples/flux_integration/neural_kernel_network/time_series.jl]:
```suggestion
```

File: test/gp/kernel.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (HamletWantToCode) [reply test/gp/kernel.jl]: Random weights has now been used to this test case :)

**Comment by HamletWantToCode** [comment]
Great, happy to see these functionalities be released :)

**Comment by willtebbutt** [comment]
Great. Will merge once I've had a final look at everything and the tests have passed!

**Comment by willtebbutt** [comment]
@HamletWantToCode looks like we need to add Flux to the `[extras]`, `[compat]` and `[targets]` in the `Project.toml`. I would suggest restricting Flux to 0.10 since we don't currently support `Tracker` for reverse-mode AD.
